File: app/backend/util/config/ByodConfigHandlerImpl.java
package backend.util.config;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.StringReader;
import java.nio.charset.StandardCharsets;
import java.util.Comparator;
import java.util.Map;
import java.util.Optional;
import java.util.TreeMap;
import java.util.zip.GZIPOutputStream;
import javax.inject.Inject;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.JsonNodeFactory;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.apache.commons.codec.digest.DigestUtils;
import org.cryptonode.jncryptor.AES256JNCryptor;
import org.cryptonode.jncryptor.CryptorException;
import org.cryptonode.jncryptor.JNCryptor;
import org.w3c.dom.Document;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;
import play.Environment;
import play.mvc.Http;
import play.mvc.Result;
import play.mvc.Results;

import backend.util.file.FileHandler;

public class ByodConfigHandlerImpl implements ByodConfigHandler {

    private static final String START_URL_PLACEHOLDER = "*** startURL ***";
    private static final String QUIT_PWD_PLACEHOLDER = "*** quitPwd ***";
    private static final String PASSWORD_ENCRYPTION = "pswd";

    private FileHandler fileHandler;
    private ConfigReader configReader;
    private Environment env;

    @Inject
    ByodConfigHandlerImpl(FileHandler fileHandler, ConfigReader configReader, Environment env) {
        this.fileHandler = fileHandler;
        this.configReader = configReader;
        this.env = env;
    }

    private String getTemplate(String hash) {
        String path = String.format("%s/conf/seb.template.plist", env.rootPath().getAbsolutePath());
        String startUrl = String.format("%s?exam=%s", configReader.getHostName(), hash);
        String template = fileHandler.read(path).replace(START_URL_PLACEHOLDER, startUrl);
        String quitPwd = DigestUtils.sha256Hex(configReader.getQuitPassword());
        return template.replace(QUIT_PWD_PLACEHOLDER, quitPwd);
    }

    private byte[] compress(byte[] data) throws IOException {
        ByteArrayOutputStream obj = new ByteArrayOutputStream();
        GZIPOutputStream gzip = new GZIPOutputStream(obj);
        gzip.write(data);
        gzip.flush();
        gzip.close();
        return obj.toByteArray();
    }

    private Optional<JsonNode> plistValueToJson(Node valueNode) {
        switch (valueNode.getNodeName()) {
            case "true":
                return Optional.of(JsonNodeFactory.instance.booleanNode(true));
            case "false":
                return Optional.of(JsonNodeFactory.instance.booleanNode(false));
            case "string":
            case "data":
                String value = valueNode.hasChildNodes() ? valueNode.getFirstChild().getNodeValue() : "";
                return Optional.of(JsonNodeFactory.instance.textNode(
                        value.trim().replaceAll("\n", "")
                ));
            case "integer":
                return Optional.of(JsonNodeFactory.instance.numberNode(
                        Integer.parseInt(valueNode.getFirstChild().getNodeValue())
                ));
            case "array":
                ArrayNode an = JsonNodeFactory.instance.arrayNode();
                for (int i = 0; i < valueNode.getChildNodes().getLength(); ++i) {
                    Node child = valueNode.getChildNodes().item(i);
                    if (child.getNodeType() == Node.ELEMENT_NODE) {
                        plistValueToJson(child).ifPresent(an::add);
                    }
                }
                return Optional.of(an);
            case "dict":
                return plistDictToJson(valueNode);
            default:
                throw new IllegalArgumentException();
        }
    }

    private Optional<JsonNode> plistDictToJson(Node dictNode) {
        if (!dictNode.hasChildNodes()) {
            return Optional.empty();
        }
        NodeList nodes = dictNode.getChildNodes();
        ObjectNode jsonNode = JsonNodeFactory.instance.objectNode();
        for (int i = 0; i < nodes.getLength(); ++i) {
            Node n = nodes.item(i);
            if (n.getNodeType() == Node.ELEMENT_NODE && n.getNodeName().equals("key")) {
                String name = n.getFirstChild().getNodeValue();
                if (!name.equals("originatorVersion")) {
                    Node valueNode = nodes.item(i + 2);
                    plistValueToJson(valueNode).ifPresent(v -> jsonNode.set(name, v));
                }
            }
        }
        return Optional.of(jsonNode);
    }

    private String jsonToSortedString(JsonNode node) throws JsonProcessingException {
        ObjectMapper om = new ObjectMapper();
        om.configure(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS, true);
        Map<String, Object> map = om.convertValue(node, new TypeReference<Map<String, Object>>() {
        });
        Map<String, Object> tm = new TreeMap<>(Comparator.comparing(String::toLowerCase));
        tm.putAll(map);
        return om.writeValueAsString(tm);
    }

    @Override
    public byte[] getExamConfig(String hash, byte[] pwd, String salt) throws IOException, CryptorException {
        String template = getTemplate(hash);
        byte[] templateGz = compress(template.getBytes(StandardCharsets.UTF_8));
        JNCryptor crypto = new AES256JNCryptor();

        // Decrypt user defined setting password
        String key = configReader.getSettingsPasswordEncryptionKey();
        byte[] saltedPwd = crypto.decryptData(pwd, key.toCharArray());
        String plainTextPwd = new String(saltedPwd, StandardCharsets.UTF_8).replace(salt, "");

        // Encrypt the config file using unencrypted password
        byte[] cipherText = crypto.encryptData(templateGz, plainTextPwd.toCharArray());
        byte[] header = PASSWORD_ENCRYPTION.getBytes(StandardCharsets.UTF_8);
        try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream()) {
            outputStream.write(header);
            outputStream.write(cipherText);
            return compress(outputStream.toByteArray());
        }
    }

    @Override
    public String getPlaintextPassword(byte[] pwd, String salt) throws CryptorException {
        // Decrypt user defined setting password
        JNCryptor crypto = new AES256JNCryptor();
        String key = configReader.getSettingsPasswordEncryptionKey();
        byte[] saltedPwd = crypto.decryptData(pwd, key.toCharArray());
        return new String(saltedPwd, StandardCharsets.UTF_8).replace(salt, "");
    }

    @Override
    public byte[] getEncryptedPassword(String pwd, String salt) throws CryptorException {
        JNCryptor crypto = new AES256JNCryptor();
        String key = configReader.getSettingsPasswordEncryptionKey();
        return crypto.encryptData((pwd + salt).getBytes(StandardCharsets.UTF_8), key.toCharArray());
    }

    @Override
    public Optional<Result> checkUserAgent(Http.RequestHeader request, String examConfigKey) {
        String protocol = request.secure() ? "https://" : "http://";
        String absoluteUrl = String.format("%s%s%s", protocol, request.host(), request.uri());

        Optional<String> oc = request.header("X-SafeExamBrowser-ConfigKeyHash");
        if (oc.isEmpty()) {
            return Optional.of(Results.unauthorized("SEB headers missing"));
        } else {
            String eckDigest = DigestUtils.sha256Hex(absoluteUrl + examConfigKey);
            if (!eckDigest.equals(oc.get())) {
                return Optional.of(Results.unauthorized("Wrong configuration key digest"));
            }
        }
        return Optional.empty();
    }


    @Override
    public String calculateConfigKey(String hash) throws IOException, ParserConfigurationException, SAXException {
        Document doc;
        try (StringReader reader = new StringReader(getTemplate(hash))) {
            InputSource src = new InputSource(reader);
            DocumentBuilder builder = DocumentBuilderFactory.newInstance().newDocumentBuilder();
            doc = builder.parse(src);
        }
        // Construct a Json-like structure out of the plist for encryption, see SEB documentation for details
        Node dictionary = doc.getDocumentElement().getElementsByTagName("dict").item(0);
        Optional<JsonNode> sebJson = plistDictToJson(dictionary);

        if (sebJson.isPresent()) {
            String ordered = jsonToSortedString(sebJson.get());
            String unescaped = ordered.replaceAll("\\\\\\\\", "\\\\");
            return DigestUtils.sha256Hex(unescaped);
        }
        throw new IOException("Failed in converting the seb.plist to json");
    }


}

File: app/backend/util/config/ByodConfigHandlerImpl.scala
package backend.util.config

import java.io.ByteArrayOutputStream
import java.net.URL
import java.nio.charset.StandardCharsets
import java.util.Optional
import java.util.zip.GZIPOutputStream

import javax.inject.Inject
import org.apache.commons.codec.digest.DigestUtils
import org.cryptonode.jncryptor.AES256JNCryptor
import play.Environment
import play.api.Logger
import play.api.libs.json._
import play.mvc.{Http, Result, Results}

import scala.compat.java8.OptionConverters._
import scala.io.Source
import scala.xml.{Node, XML}

object ByodConfigHandlerImpl {
  private val StartUrlPlaceholder      = "*** startURL ***"
  private val QuitPwdPlaceholder       = "*** quitPwd ***"
  private val QuitLinkPlaceholder      = "*** quitLink ***"
  private val AdminPwdPlaceholder      = "*** adminPwd ***"
  private val AllowQuittingPlaceholder = "<!-- allowQuit /-->"
  private val PasswordEncryption       = "pswd"
  private val ConfigKeyHeader          = "X-SafeExamBrowser-ConfigKeyHash"
  private val IgnoredKeys              = Seq("originatorVersion")
}
class ByodConfigHandlerImpl @Inject()(configReader: ConfigReader, env: Environment)
    extends ByodConfigHandler {

  import ByodConfigHandlerImpl._
  private val logger        = Logger(this.getClass).logger
  private val crypto        = new AES256JNCryptor
  private val encryptionKey = configReader.getSettingsPasswordEncryptionKey

  /* FIXME: have Apache provide us with X-Forwarded-Proto header so we can resolve this automatically */
  private val protocol = new URL(configReader.getHostName).getProtocol

  private def getTemplate(hash: String): Node = {
    val path          = s"${env.rootPath.getAbsolutePath}/conf/seb.template.plist"
    val startUrl      = s"${configReader.getHostName}?exam=$hash"
    val quitLink      = configReader.getQuitExaminationLink
    val adminPwd      = DigestUtils.sha256Hex(configReader.getExaminationAdminPassword)
    val quitPwdPlain  = configReader.getQuitPassword
    val quitPwd       = DigestUtils.sha256Hex(quitPwdPlain)
    val allowQuitting = if (quitPwdPlain.isEmpty) "<false/>" else "<true/>"
    val source        = Source.fromFile(path)
    val template = source.mkString
      .replace(StartUrlPlaceholder, startUrl)
      .replace(QuitLinkPlaceholder, quitLink)
      .replace(AdminPwdPlaceholder, adminPwd)
      .replace(QuitPwdPlaceholder, quitPwd)
      .replace(AllowQuittingPlaceholder, allowQuitting)
    source.close
    XML.loadString(template)
  }

  private def compress(data: Array[Byte]): Array[Byte] = {
    val os   = new ByteArrayOutputStream
    val gzip = new GZIPOutputStream(os)
    gzip.write(data)
    gzip.flush()
    gzip.close()
    os.toByteArray
  }

  private def compressWithHeader(data: Array[Byte]): Array[Byte] = {
    val header = PasswordEncryption.getBytes(StandardCharsets.UTF_8)
    val os     = new ByteArrayOutputStream()
    os.write(header)
    os.write(data)
    os.close()
    compress(os.toByteArray)
  }

  private def nodeToJson(node: Node): Option[JsValue] = node.label match {
    case "string" | "data" =>
      val text = if (node.child.nonEmpty) node.child.head.text else ""
      Some(JsString(text.trim.filterNot(_ == '\n')))
    case "true"    => Some(JsBoolean(true))
    case "false"   => Some(JsBoolean(false))
    case "integer" => Some(JsNumber(node.child.head.text.toInt))
    case "array"   => Some(JsArray(node.child.flatMap(nodeToJson)))
    case "dict"    => dictToJson(node)
    case _         => throw new NoSuchElementException
  }

  private def dictToJson(dict: Node): Option[JsValue] = dict.child match {
    case Seq() => None
    case children =>
      val json: Seq[(String, JsValue)] = children
        .grouped(2)
        .map(c => (c.head.text, c.last))
        .filterNot(c => IgnoredKeys.contains(c._1))
        .map(n => (n._1, nodeToJson(n._2)))
        .filterNot(_._2.isEmpty)
        .map(n => n._1 -> n._2.get)
        .toSeq
        .sortBy(_._1.toLowerCase)
      Some(JsObject(json))
  }

  override def getExamConfig(hash: String, pwd: Array[Byte], salt: String): Array[Byte] = {
    val template   = getTemplate(hash)
    val templateGz = compress(template.toString.getBytes(StandardCharsets.UTF_8))
    // Decrypt user defined setting password
    val plaintextPwd = getPlaintextPassword(pwd, salt)
    // Encrypt the config file using unencrypted password
    val cipherText = crypto.encryptData(templateGz, plaintextPwd.toCharArray)
    compressWithHeader(cipherText)
  }

  override def getPlaintextPassword(pwd: Array[Byte], salt: String): String = {
    val saltedPwd = crypto.decryptData(pwd, encryptionKey.toCharArray)
    new String(saltedPwd, StandardCharsets.UTF_8).replace(salt, "")
  }

  override def getEncryptedPassword(pwd: String, salt: String): Array[Byte] =
    crypto.encryptData((pwd + salt).getBytes(StandardCharsets.UTF_8), encryptionKey.toCharArray)

  override def checkUserAgent(request: Http.RequestHeader, configKey: String): Optional[Result] =
    request.header(ConfigKeyHeader).asScala match {
      case None => Some(Results.unauthorized("SEB headers missing")).asJava
      case Some(digest) =>
        val absoluteUrl = s"$protocol://${request.host}${request.uri}"
        DigestUtils.sha256Hex(absoluteUrl + configKey) match {
          case sha if sha == digest => None.asJava
          case sha =>
            logger.warn(
              "Config key mismatch for URL {} and exam config key {}. Digest received: {}",
              absoluteUrl,
              configKey,
              sha)
            Some(Results.unauthorized("Wrong configuration key digest")).asJava
        }
    }

  override def calculateConfigKey(hash: String): String = {
    val plist: Node = getTemplate(hash)
    // Construct a Json-like structure out of .plist and create a digest over it
    // See SEB documentation for details
    dictToJson((plist \ "dict").head) match {
      case Some(json) => DigestUtils.sha256Hex(json.toString.replaceAll("\\\\\\\\", "\\\\"))
      case None       => throw new NoSuchElementException
    }
  }
}
